tensorflow_addons/optimizers/discriminative_layer_training.py

```diff
@@ -0,0 +1,255 @@
+        """
+
+        mult = self._get_lr_mult(layer)
+        layers = self._get_layers(layer)
+
+        if len(layers) > 0:
+            for sublayer in layers:
+
+                # we generally want to propagate the lr mult to the lower layers
+                if propagate_lr_mult_to_sub_layers:
+                    self._assign_lr_mult(sublayer, mult)
+
+                # recursively iterate through the nested layers
+                for nested_sublayer in self._get_lowest_layers(sublayer):
+                    yield nested_sublayer
+
+        else:
+            yield layer
+
+    def _apply_lr_mult_to_var(self, layer):
+        """Helper method to apply the lr mult to the trainable variables of a layer
+        """
+
+        lr_mult = self._get_lr_mult(layer)
+
+        for var in layer.trainable_variables:
+            var.lr_mult = tf.convert_to_tensor(lr_mult, tf.float32)  # 0D tensor
+            var.lr_mult_value = (
+                lr_mult  # easier to check vars lr mult in graph and eager
+            )
+
+    def _check_for_lr_mult(self, layer, verbose=True, propagate=True):
+        """Identify which layers have an LR mult not equal to 1
+        """
+
+        layers_with_lr_mult = []
+
+        for sub_layer in self._get_lowest_layers(
+            layer, propagate_lr_mult_to_sub_layers=propagate
+        ):
+            lr_mult = self._get_lr_mult(sub_layer)
+            if lr_mult != 1.0:
+                layers_with_lr_mult.append(sub_layer)
+                if verbose:
+                    logging.info("layer %s lr_mult : %f" % (sub_layer.name, lr_mult))
+
+        return layers_with_lr_mult
+
+    def _compute_params(self, var_list):
+        """helps compute params to provide a summary that aligns with model.summary()
+        """
+        return np.sum([np.prod(list(var.shape)) for var in var_list])
+
+    def _prepare_model(self, model, verbose=True):
+        """Prepares a built model for disc training
+        """
+
+        layers_with_lr_mult = self._check_for_lr_mult(model, verbose=verbose)
+        if len(layers_with_lr_mult) == 0:
+
+            logging.warning(
+                "Discriminative Layer Training requires an lr_mult attribute on at least one layer"
+            )
+
+            logging.warning(
+                "The assigned lr_mult must not be equal to 1. eg: model.layers[0].lr_mult = 0.01"
+            )
+
+        for layer in self._get_lowest_layers(model):
+            self._apply_lr_mult_to_var(layer)
+
+        vars_with_lr_mult = [
+            var for var in model.trainable_variables if var.lr_mult_value != 1.0
+        ]
+
+        if verbose:
+            logging.info(
+                "%i params of %i will learn at a different rate"
+                % (
+                    self._compute_params(vars_with_lr_mult),
+                    self._compute_params(model.trainable_variables),
+                )
+            )
+
+
+class DiscriminativeWrapper(tf.keras.optimizers.Optimizer):
```

Comment:
This name is lacking IMO, `DiscriminativeLayerOptimizer`?

Reply:
That sounds good. I'll use DiscriminativeLayerOptimizer